operator: Add LB-IPAM (#21764)
cilium-team/cilium: Documentation/network/lb-ipam.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _lb_ipam:

********************************************
LoadBalancer IP Address Management (LB IPAM)
********************************************

LB IPAM is a feature that allows Cilium to assign IP addresses to Services of
type ``LoadBalancer``. This functionally is usually left up to a cloud provider,
however, when deploying in a private cloud environment, these facilities are not
always available.

> LB IPAM work in conjunction with features like the :ref:`bgp_control_plane`. Where

Reviewer: ```suggestion
LB IPAM works in conjunction with features like the :ref:`bgp_control_plane`. Where
```

LB IPAM is responsible for allocation and assigning of IPs to Service objects and
other features are responsible for load balancing and/or advertisement of these
IPs. 

> LB IPAM is implicitly enabled when dependant features are enabled.

Reviewer: Which are they?

Pools
#####

LB IPAM has the notion of IP Pools which the administrator can create to tell 
Cilium which IP ranges can be used to allocate IPs from.

> A basic IP Pools with both an IPv4 and IPv6 range looks like.

Reviewer: ```suggestion
A basic IP Pools with both an IPv4 and IPv6 range looks like this:
```

.. code-block:: yaml

    apiVersion: "cilium.io/v2alpha1"
    kind: CiliumLoadBalancerIPPool
    metadata:
      name: "blue-pool"
    spec:
      cidrs:
      - cidr: "10.0.10.0/24"
      - cidr: "2004::0/64"

After adding the pool to the cluster, it appears like so.

.. code-block:: shell-session

    $ kubectl get ippools                           
    NAME        DISABLED   CONFLICTING   IPS AVAILABLE   AGE
    blue-pool   false      false         65788           2s

.. note::

>     The amount of available IPs in the pool is less than the actual sum of all

Reviewer: ```suggestion
    The amount of available IPs in the pool is lower than the actual sum of all
```

    usable IPs in the CIDRs because the allocation logic is limited to 65536 IPs

>     per CIDR. CIDRs containing more than 65536 IPs can be broken up into multiple

Reviewer: Not sure, but I'd rather say “broken down”?

```suggestion
    per CIDR. CIDRs containing more than 65536 IPs can be broken down into multiple
```

Author: I don't know honestly. If I say "breaking up a problem" I mean as in smaller pieces just like "breaking up a group". but "breaking down a problem" is also solving in smaller parts, like breaking "down a building". Lets go with "down" for now.

The pool will allocate to any service if no service selector is specified.

.. code-block:: yaml

    apiVersion: "cilium.io/v2alpha1"
    kind: CiliumLoadBalancerIPPool
    metadata:
      name: "blue-pool"
    spec:
      cidrs:
      - cidr: "20.0.10.0/24"
      serviceSelector:
        matchExpressions:
          - {key: color, operator: In, values: [blue, cyan]}
    ---
    apiVersion: "cilium.io/v2alpha1"
    kind: CiliumLoadBalancerIPPool
    metadata:
      name: "red-pool"
    spec:
      cidrs:
      - cidr: "20.0.10.0/24"
      serviceSelector:
        matchLabels:
          color: red

There are a few special purpose selector fields which don't match on labels but
instead on other metadata like ``.meta.name`` or ``.meta.namespace``.

=============================== ===================
Selector                        Field
------------------------------- -------------------
io.kubernetes.service.namespace ``.meta.namespace``
io.kubernetes.service.name      ``.meta.name``
=============================== ===================

> .. code-block:: yaml

Reviewer: ```suggestion
For example:

.. code-block:: yaml
```

    apiVersion: "cilium.io/v2alpha1"
    kind: CiliumLoadBalancerIPPool
    metadata:
      name: "blue-pool"
    spec:
      cidrs:
      - cidr: "20.0.10.0/24"
      serviceSelector:
        matchLabels:
          "io.kubernetes.service.namespace": "tenant-a"

Conflicts
---------

IP Pools are not allowed to have overlapping CIDRs. When an administrator does
create pools which overlap, a soft error is caused. The last added pool will be
marked as ``Conflicting`` and no further allocation will happen from that pool.

> Therefor administrators should always check the status of all pools after making

Reviewer: ```suggestion
Therefore, administrators should always check the status of all pools after making
```

modifications.

> If we for example add 2 pools(``blue-pool`` and ``red-pool``) both with the same

Reviewer: I won't let go regarding parenthesis :)

```suggestion
For example, if we add 2 pools (``blue-pool`` and ``red-pool``) both with the same
```

Author: Sigh. You shouldn't. Just just to writing `if(`, `for(` and `somefunc(`. Doesn't translate that will into non-code formatting.

    Pool conflicts since CIDR '20.0.10.0/24' overlaps CIDR '20.0.10.0/24' from IP Pool 'blue-pool'

or

.. code-block:: shell-session

    $ kubectl describe ippools/red-pool
    Name:         red-pool
    #[...]
    Status:
      #[...]
      Conflict Reason:  Pool conflicts since CIDR '20.0.10.0/24' overlaps CIDR '20.0.10.0/24' from IP Pool 'blue-pool'
      Conflicting:      true
      #[...]

Disabling a Pool
-----------------

IP Pools can be disabled. Disabling a pool will stop LB IPAM from allocating
new IPs from the pool, but doesn't remove existing allocations. This allows
an administrator to slowly drain pool or reserve a pool for future use.

.. code-block:: yaml

    apiVersion: "cilium.io/v2alpha1"
    kind: CiliumLoadBalancerIPPool
    metadata:
      name: "blue-pool"
    spec:
      cidrs:
      - cidr: "20.0.10.0/24"
      disabled: true

.. code-block:: shell-session

    $ kubectl get ippools          
    NAME        DISABLED   CONFLICTING   IPS AVAILABLE   AGE
    blue-pool   true       false         254             41m

Status
------

The IP Pool's status contains additional counts which can be used to monitor

> the amount of used and available IPs. A machine parsable output can be gotten like so.

Reviewer: I'm not familiar with “gotten”, but that's maybe OK. I'm nitpicking here.

```suggestion
the amount of used and available IPs. A machine parsable output can be obtained like so.
```

Author: `gotten` is more American English(I tend to use both since it is not my primary language), `obtained` does sound better, will go with that.

Services can use the ``.spec.ipFamilyPolicy`` and ``.spec.ipFamilies`` fields to change
the requested IPs.

If ``.spec.ipFamilyPolicy`` isn't specified, ``SingleStack`` mode is assumed. 
If both IPv4 and IPv6 are enabled in ``SingleStack`` mode, an IPv4 address is allocated.

If ``.spec.ipFamilyPolicy`` is set to ``PreferDualStack``, LB IPAM will attempt to allocate 
both an IPv4 and IPv6 address if both are enabled on the cluster. If only IPv4 or only IPv6 is
enabled on the cluster, the service is still considered "satisfied".

If ``.spec.ipFamilyPolicy`` is set to ``RequireDualStack`` LB IPAM will attempt to allocate
both an IPv4 and IPv6 address. The service is considered "unsatisfied" If IPv4 
or IPv6 is disabled on the cluster.

The order of ``.spec.ipFamilies`` has no effect on LB IPAM but is significant for cluster IP
allocation which isn't handled by LB IPAM.

LoadBalancerClass
-----------------

Kubernetes >= v1.24 supports `multiple load balancers <https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class>`_ 
in the same cluster. Picking between load balancers is done with the ``.spec.loadBalancerClass`` field. 

> When LB IPAM is enabled it assumed that it is supposed to allocate and assign IPs for services with 

Reviewer: ```suggestion
When LB IPAM is enabled it allocates and assigns IPs for services with 
```

no load balancer class set.

> LB IPAM only does IP allocation and doesn't provide load balancing services by itself. Therefor

Reviewer: ```suggestion
LB IPAM only does IP allocation and doesn't provide load balancing services by itself. Therefore,
```

users should pick one of the following Cilium load balancer classes, all of which use LB IPAM
for allocation (if the feature is enabled):

=============================== ========================
loadBalancerClass               Feature
------------------------------- ------------------------
``io.cilium/bgp-control-plane`` :ref:`bgp_control_plane`
=============================== ========================

If the ``.spec.loadBalancerClass`` is set to a class which isn't handled by LB IPAM, then LB IPAM
will ignore the service entirely, not even setting a condition in the status. 

Requesting IPs
--------------

> Services can request specific IPs. The legacy wait of doing so is via ``.spec.loadBalancerIP``

Reviewer: ```suggestion
Services can request specific IPs. The legacy way of doing so is via ``.spec.loadBalancerIP``
```

which takes a single IP address. This method has been deprecated in k8s v1.24 but is supported
until its future removal.

The new way of requesting specific IPs is to use annotations, ``io.cilium/lb-ipam-ips`` in the case

> of Cilium LB IPAM. This annotation takes a comma separated list of IP addresses, allowing for

Reviewer: ```suggestion
of Cilium LB IPAM. This annotation takes a comma-separated list of IP addresses, allowing for
```

multiple IPs to be requested at once.

The service selector of the IP Pool still applies, requested IPs will not be allocated or assigned

> if the services doesn't match the pool's selector.

Reviewer: ```suggestion
if the services don't match the pool's selector.
```

or IPv6 is disabled on the cluster.

The order of ``.spec.ipFamilies`` has no effect on LB IPAM but is significant for cluster IP
allocation which isn't handled by LB IPAM.

LoadBalancerClass
-----------------

Kubernetes >= v1.24 supports `multiple load balancers <https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class>`_ 
in the same cluster. Picking between load balancers is done with the ``.spec.loadBalancerClass`` field. 
When LB IPAM is enabled it allocates and assigns IPs for services with 
no load balancer class set.

LB IPAM only does IP allocation and doesn't provide load balancing services by itself. Therefore,
users should pick one of the following Cilium load balancer classes, all of which use LB IPAM
for allocation (if the feature is enabled):

=============================== ========================
loadBalancerClass               Feature
------------------------------- ------------------------
``io.cilium/bgp-control-plane`` :ref:`bgp_control_plane`
=============================== ========================

If the ``.spec.loadBalancerClass`` is set to a class which isn't handled by LB IPAM, then LB IPAM

> will ignore the service entirely, not even setting a condition in the status. 

Reviewer 2: nit (to make it a bit more obvious that other components outside of Cilium might handle it):
```suggestion
If the ``.spec.loadBalancerClass`` is set to a class which isn't handled by Cilium's LB IPAM, then Cilium's LB IPAM
will ignore the service entirely, not even setting a condition in the status. 
```

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _lb_ipam:

********************************************
LoadBalancer IP Address Management (LB IPAM)
********************************************

LB IPAM is a feature that allows Cilium to assign IP addresses to Services of

> type ``LoadBalancer``. This functionally is usually left up to a cloud provider,

Reviewer 3: ```suggestion
type ``LoadBalancer``. This functionality is usually left up to a cloud provider,
```
?
